core/app/models/spree/discounts/shipment_updater.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Spree
+  module Discounts
+    class ShipmentUpdater
+      attr_reader :promotions
+
+      def initialize(promotions:)
+        @promotions = promotions
+      end
+
+      def call(shipment)
+        shipment.discounts = promotions.select do |promotion|
+          promotion.shipment_eligible?(shipment)
+        end.flat_map do |promotion|
+          promotion.actions.select do |action|
+            action.discountable_class == Spree::Shipment
+          end.map do |action|
+            action.discount(shipment)
+          end
+        end
+
+        chosen_discounts = Spree::Config.discount_chooser_class.new(shipment).call(discounts)
+        shipment.promo_total = chosen_discounts.sum(&:amount)
```

Comment:
Should the `order_updater` do this instead?

Reply:
yes